File: app/src/domain/calculations/remainingBudgetRate.ts
/**
 * 残予算必要達成率
 *
 * 残りの予算を、残り期間の計画予算に対して何％必要かを算出する。
 *
 * 式: (月間予算 - 累計実績) ÷ 残期間予算 × 100
 *
 * - 残期間予算 = Σ budgetDaily[elapsedDays+1 .. daysInMonth]
 * - 100% = 計画通りのペース、100%超 = 巻き返しが必要
 */
import { z } from 'zod'
import { safeDivide } from './utils'

export interface RemainingBudgetRateInput {
  /** 月間予算 */
  readonly budget: number
  /** 累計実績売上 */
  readonly totalSales: number
  /** 日別予算マップ (day → amount) */
  readonly budgetDaily: ReadonlyMap<number, number>
  /** 経過日数 */
  readonly elapsedDays: number
  /** 月日数 */
  readonly daysInMonth: number
}

export const RemainingBudgetRateInputSchema = z.object({
  budget: z.number(),
  totalSales: z.number(),
  budgetDaily: z.map(z.number(), z.number()),
  elapsedDays: z.number(),
  daysInMonth: z.number(),
})

/**
 * 残予算必要達成率を算出する（%値: 100 = 計画通り）
 *
 * @returns 達成率（%表示済み）。残期間予算が 0 なら 0 を返す。
 */
export function calculateRemainingBudgetRate(input: RemainingBudgetRateInput): number {
  const { budget, totalSales, budgetDaily, elapsedDays, daysInMonth } = input

  // 残予算額 = 月間予算 − 累計実績
  const remainingTarget = budget - totalSales

  // 残期間予算 = Σ budgetDaily[elapsedDays+1 .. daysInMonth]
  let remainingPeriodBudget = 0
  for (let d = elapsedDays + 1; d <= daysInMonth; d++) {
    remainingPeriodBudget += budgetDaily.get(d) ?? 0
  }

  return safeDivide(remainingTarget, remainingPeriodBudget, 0) * 100
}

File: wasm/remaining-budget-rate/src/remaining_budget_rate.rs
/// Remaining budget achievement rate calculation.
///
/// Formula: (monthly_budget - total_sales) / remaining_period_budget * 100
/// Where remaining_period_budget = Σ budget_daily[elapsed_days+1 .. days_in_month]
///
/// Returns 0 if remaining period budget is 0.
/// 100 = on plan, >100 = catch-up needed.
///
/// @contractId BIZ-008
/// @semanticClass business
/// @methodFamily budget

use crate::utils::safe_divide;

/// Percentage conversion multiplier (matches TS `* 100` in remainingBudgetRate.ts)
const PERCENT_MULTIPLIER: f64 = 100.0;

/// Calculate remaining budget rate.
///
/// `budget_daily` is a flat array where index i represents day (i+1)'s budget amount.
/// Array length should be >= days_in_month.
pub fn calculate_remaining_budget_rate(
    budget: f64,
    total_sales: f64,
    budget_daily: &[f64],
    elapsed_days: usize,
    days_in_month: usize,
) -> f64 {
    let remaining_target = budget - total_sales;

    let mut remaining_period_budget = 0.0;
    for d in elapsed_days..days_in_month {
        if d < budget_daily.len() {
            remaining_period_budget += budget_daily[d];
        }
    }

    safe_divide(remaining_target, remaining_period_budget, 0.0) * PERCENT_MULTIPLIER
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn on_plan() {
        // 30 days, 10 elapsed, budget 300, sales 100
        // Remaining target = 200, remaining budget = 20 * 10 = 200
        let daily: Vec<f64> = vec![10.0; 30];
        let result = calculate_remaining_budget_rate(300.0, 100.0, &daily, 10, 30);
        assert_eq!(result, 100.0);
    }

    #[test]
    fn catch_up_needed() {
        // Behind plan: sales only 50, need 250 out of remaining 200
        let daily: Vec<f64> = vec![10.0; 30];
        let result = calculate_remaining_budget_rate(300.0, 50.0, &daily, 10, 30);
        assert_eq!(result, 125.0);
    }

    #[test]
    fn ahead_of_plan() {
        // Ahead: sales 200, only need 100 out of remaining 200
        let daily: Vec<f64> = vec![10.0; 30];
        let result = calculate_remaining_budget_rate(300.0, 200.0, &daily, 10, 30);
        assert_eq!(result, 50.0);
    }

    #[test]
    fn zero_remaining_budget() {
        // All days elapsed → remaining budget = 0 → returns 0
        let daily: Vec<f64> = vec![10.0; 30];
        let result = calculate_remaining_budget_rate(300.0, 100.0, &daily, 30, 30);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn empty_daily() {
        let daily: Vec<f64> = vec![];
        let result = calculate_remaining_budget_rate(300.0, 100.0, &daily, 0, 30);
        assert_eq!(result, 0.0);
    }
}
